dask/array/core.py

```diff
@@ -1497,6 +1497,16 @@ def __xor__(self, other):
     def __rxor__(self, other):
         return elemwise(operator.xor, other, self)
 
+    def __matmul__(self, other):
+        return tensordot(self, other, axes=((self.ndim - 1,),
+                                            (other.ndim - 2,)))
+
+    def __rmatmul__(self, other):
+        if not hasattr(other, 'ndim'):
+            other = np.asarray(other)  # account for array-like on LHS
+        return tensordot(other, self, axes=((other.ndim - 1,),
+                                            (self.ndim - 2,)))
```

Comment:
@mrocklin sure, and this is very cool. However, those containers will have an `ndim` attribute and thus not get coerced, right? =)